### DISCUSSION TITLE
recommandation on modeling database in pocketbase

### DISCUSSION THREAD

**Discussion by sanathusk** [pocketbase/pocketbase#1776] [Q&A] [answered]
I am try to build a simple Todo application using Pocketbase where each user can Create/Read/Update/Delete todos that is private the user.I needed recommendation around the correct way to design the use case.
Initially i though of adding fields a list todo field inside the users collection.What should the type the of todos field? Should it be a select or a relation? I have created another `todos` collection with fields user, description. Now should the user by of type relation ? I have exposure to relational database and often think in terms of relations and foreign key. Since pocketbase db is NoSQL database, shouldn't  I  be allowed to nest the todo collection inside my user collection? Should todos be a root collection? I am frontend developer and my db skills are not that great. Any suggestion would be for great help.

**Comment by ganigeorgiev** [comment, marked as answer]
>  Since pocketbase db is NoSQL database, shouldn't I be allowed to nest the todo collection inside my user collection?

This is incorrect. PocketBase internally uses SQLite and under the hood all collections are plain SQL tables. We don't use directly foreign keys because relations are "dynamic" (this will change in the future), but referencing a relation is via the primary key (`id`) of the related collection record, so on the surface is the same. More information about the collections you can find in https://pocketbase.io/docs/collections/.

> Should todos be a root collection? 

I'm not sure that I understand what you mean by "root". At the moment all collections can have only "flat" structure, aka. there is no nesting. You can have a `multi`-relation field that contains references to the todo items but the `todos` record must have their own collection.

---

About your use case - there are several ways you can accomplish this depending on your desired workflow and data management. Here is an example of the simplest use case (aka. single todos list for each user):

1. Create a `todos` collection with the folling fields:
   - `user` (nonempty `relation` field pointing to the `users` collection and with "Max select = 1")
   - `title` (`text` field)
   - ... any other field

2. For the `todos` API rules set:
    ```js
    // List, View, Update and Delete
    user = @request.auth.id
    
    // Create
    @request.data.user = @request.auth.id
    ```

To create a todo item:
```js
await pb.collection("users").authWithPassword("test@example.com", "1234567890")

const todo = await pb.collection("todos").create({
    "title": "Todo 1",
    "user": pb.authStore.model.id,
})
```

To fetch all user's todos:
```js
await pb.collection("users").authWithPassword("test@example.com", "1234567890")

const todos = await pb.collection("todos").getFullList()
```

---

If you want to allow users to have multiple todos list each with many todo items, then you can create a `todoLists` and `todoItems` collections.

1. `todoLists`:
   
   - `title` (`text` field)
   - `user` (nonempty `relation` field pointing to the `users` collection and with "Max select = 1")
   - ... any other field


    1.1. For the `todoLists` API rules set:
    ```js
    // List, View, Update and Delete
    user = @request.auth.id
    
    // Create
    @request.data.user = @request.auth.id
    ```

2. `todoItems`:
   - `list` (nonempty `relation` field pointing to the `todoLists` collection and with "Max select = 1")
   - `message` (`text` field)
   - ... any other field


    2.1. For the `todoItems` API rules set:
    ```js
    // List, View, Update and Delete
    list.user = @request.auth.id
    
    // Create
    @request.data.list.user = @request.auth.id
    ```

To create a todo list with 2 todo items:
```js
await pb.collection("users").authWithPassword("test@example.com", "1234567890")

const list = await pb.collection("todoLists").create({
    "title": "Todo list 1",
    "user": pb.authStore.model.id,
    ...
})

const item1 = await pb.collection("todoItems").create({
    "list": list.id,
    ...
})

const item2 = await pb.collection("todoItems").create({
    "list": list.id,
    ...
})
```

To fetch a single (_or multiple if `getList`/`getFullList` are used_) user's todo list with all todo items expanded:
```js
await pb.collection("users").authWithPassword("test@example.com", "1234567890")

const listWithTodoItems = await pb.collection("todoList").getOne("LIST_ID", { expand: "todoItems(list)" })
```

_**Edit:** I've updated the rules to show the individual actions constraints_

**Reply by sanathusk** [reply]
Thank you @ganigeorgiev for such an elaborate answer 🙏
